tools/oesign/oedump.c

```diff
@@ -0,0 +1,249 @@
+{
+    int rc = -1;
+    visit_sym_data_t* data = (visit_sym_data_t*)data_;
+    const elf64_shdr_t* shdr = data->shdr;
+    const char* name;
+
+    data->result = OE_UNEXPECTED;
+
+    /* Skip symbol if not a function */
+    if ((sym->st_info & 0x0F) != STT_FUNC)
+    {
+        rc = 0;
+        goto done;
+    }
+
+    /* Skip symbol if not in the ".ecall" section */
+    if (sym->st_value < shdr->sh_addr ||
+        sym->st_value + sym->st_size > shdr->sh_addr + shdr->sh_size)
+    {
+        rc = 0;
+        goto done;
+    }
+
+    /* Skip null names */
+    if (!(name = elf64_get_string_from_dynstr(data->elf, sym->st_name)))
+    {
+        rc = 0;
+        goto done;
+    }
+
+    /* Dump the ECALL name */
+    printf("%s (%016llx)\n", name, OE_LLX(sym->st_value));
+
+    rc = 0;
+
+done:
+    return rc;
+}
+
+void dump_ecall_section(elf64_t* elf)
+{
+    elf64_shdr_t shdr;
+
+    printf("=== ECALLs:\n");
+
+    /* Find the .ecall section */
+    if (elf64_find_section_header(elf, ".ecall", &shdr) != 0)
+    {
+        err("missing .ecall section");
+        return;
+    }
+
+    /* Dump all the ECALLs */
+    {
+        visit_sym_data_t data;
+        data.elf = elf;
+        data.shdr = &shdr;
+
+        if (elf64_visit_symbols(elf, _visit_sym, &data) != 0)
+        {
+            err("failed to find ECALLs in .ecall section");
+            return;
+        }
+    }
+
+    printf("\n");
+}
+
+void check_global(elf64_t* elf, const char* name)
+{
+    elf64_sym_t sym;
+
+    if (elf64_find_dynamic_symbol_by_name(elf, name, &sym) != 0)
+    {
+        err("failed to find required symbol: %s\n", name);
+        return;
+    }
+
+    printf("%s (%016llx)\n", name, OE_LLX(sym.st_value));
+}
+
+void check_globals(elf64_t* elf)
+{
+    printf("=== Globals:\n");
+
+    check_global(elf, "oe_num_pages");
+    check_global(elf, "oe_virtual_base_addr");
+    check_global(elf, "oe_base_reloc_page");
+    check_global(elf, "oe_num_reloc_pages");
+    check_global(elf, "oe_base_ecall_page");
+    check_global(elf, "oe_num_ecall_pages");
+    check_global(elf, "oe_base_heap_page");
+    check_global(elf, "oe_num_heap_pages");
+
+    printf("\n");
+}
+
+void oedump(const char* arg0, const char* enc_bin, const char* option)
+{
+    elf64_t elf;
+    oe_sgx_enclave_properties_t props;
+
+    if ((strcmp(option, "--oedump") != 0))
+    {
+        fprintf(stderr, "Usage: %s EnclaveImage --oedump\n", arg0);
+        return;
+    }
+
+    /* Load the ELF-64 object */
+    if (elf64_load(enc_bin, &elf) != 0)
+    {
+        fprintf(stderr, "failed to load %s\n", enc_bin);
+        goto done;
+    }
+
+    /* Load the SGX enclave properties */
+    if (oe_sgx_load_properties(&elf, OE_INFO_SECTION_NAME, &props) != OE_OK)
+    {
+        err("failed to load SGX enclave properties from %s section",
+            OE_INFO_SECTION_NAME);
+    }
+
+    printf("\n");
+
+    /* Dump the entry point */
+    dump_entry_point(&elf);
```

Comment:
@mikbras Is this valuable to keep? We use `_start` now, and we don't expect there to be an explicit flag on the build line to enable this, so I'm not sure if dumping this is useful anymore. This will also end up being something different in PE so I'd like to remove things that have to be supported in different ways on different systems.